### DISCUSSION TITLE
Permissions

### DISCUSSION THREAD

**Discussion by dtcu8ey6p41w** [RealRaven2000/FiltaQuilla#305] [Q&A]
Hi there, 

First of al thanks for making nice and useful add-ons, life would be less nice without them ;) 

When I look at the permissions FiltaQuilla needs when I want to install the add-ons I see: 

This add-on can: Have full, unrestricted access to Thunderbird, and your computer

and

![image](https://github.com/user-attachments/assets/453f13c4-a523-4ff0-a96c-719eabeb7c11)

I try to be careful with what (add-ons/ apps) I install on my machines and I understand that to use certain functionalities you need certain permissions but when I read the permissions above it feels kind of wild "full, unrestricted access to Thunderbird, and your computer".

With my limited knowledge of coding it's hard for me to oversee what that means. I also read negative stories about 'trusted' add-ons that are being sold over time to different type of companies that start abusing permissions. 

When I look at how active you are, how much you contribute to the TB community etc. I would think it's 'save' to give the add-on access but it's still a guess for me what the potential (future) threads are. Full access to TB means also to all email, all calendars and tasks. And full access to my computer?! I have not even an idea what that means, to all my files? 

I hope you can explain a bit what I can expect and what this scary sounding permissions mean. 

Thanks! L

**Comment by RealRaven2000** [comment]
After **Thunderbird 78 dropped legacy** Add-on support, advanced Add-ons like FiltaQuilla use a mixture of official mail APIs and our own "experimental APIs", which allow us access to the backend code of Thunderbird. This programming model has full access to the underlying programming interfaces, which means all features and capabilities of Thunderbird. This also necessitates adaptations in the Add-on in case Thunderbird changes code, variables or UI implementation internally.

At the moment there is no API for accessing or modifying filters so this is the only way this can be implemented. All old Add-ons had permission to script the scriptable parts of Mozilla apps (Firefox / Thunderbird) in order to modify / augment their behavior. That's why such Add-ons have to be fully code reviewed by community members, the complete source code must be checked for security risks such as leaking data etc.

With the newer permissions-driven web extensions which are based on APIs this responsibility is passed on to the users: they have to allow the Add-on access to areas of functionality, such as address book access, read emails, modify emails etc.

FiltaQuilla can add new filter actions and search terms, and under actions you have functions like saving attachments to file (file access is generally not allowed to any web extension) and running external programs (if you choose you can run any other process via a FiltaQuilla action, but you have to determine that as a user when you create the filter). This stuff will probably never possible with a web extension that is permission driven as it deemed "too risky" from the perspective of a web browser (Thunderbird inherited the web extension Add-ons model from Firefox). 

That's why Thunderbird (which is essentially a database fat client) added the model of "experimental" Add-ons that still allowed us to use the old paradigm - script what you can script. By the way you are not safe from abuse if you just go with web extensions, if you give "address book permission" to a web extension, how do you know what it will exactly do with the address book data? You will still have to put in a level of trust to install it, and the source code of web extensions is not reviewed by human eyes. So in that sense I would judge (even with the scary permissions warning) that our "experimental" Add-ons (or as you might call them "legacy" add-ons) are just as safe or unsafe as web extensions. And I am publishing the complete source code here on GIthub, there is nothing hidden and it doesn't use any third party frameworks.

In the meantime I am awaiting the creation of a filters API which will at least create new permissions for filters, but I wouldn't hold my breath that anybody can actually create something like FiltaQuilla with this because the filtering mechanism is written in C++ via a service, and I do not see how this could be scripted via an API.

**Reply by dtcu8ey6p41w** [reply]
Thank you for the comprehensive response. To be clear, it's not a lack of trust in you or your work that makes me question the permissions, it's the extreme data hunger, endless stories about abuse, hacks, data breaches etc. that makes me careful.
